core/services/health/health.go

```diff
@@ -0,0 +1,219 @@
+)
+
+// Types requiring health checks should implement the Checkable interface.
+type Checkable interface {
+	// Checkables should return nil if ready, or an error message otherwise.
+	Ready() error
+	// Checkables should return nil if healthy, or an error message otherwise.
+	Healthy() error
+}
+
+//go:generate mockery --name Checker --output ../../internal/mocks/ --case=underscore
+type (
+	// Checker provides a service which can be probed for system health.
+	Checker interface {
+		// Register a service for health checks.
+		Register(name string, service Checkable) error
+		// Unregister a service.
+		Unregister(name string) error
+		// IsReady returns the current readiness of the system.
+		// A system is considered ready if all checks are passing (no errors)
+		IsReady() (ready bool, errors map[string]error)
+		// IsHealthy returns the current health of the system.
+		// A system is considered healthy if all checks are passing (no errors)
+		IsHealthy() (healthy bool, errors map[string]error)
+
+		Start() error
+		Close() error
+	}
+
+	checker struct {
+		srvMutex   sync.RWMutex
+		services   map[string]Checkable
+		stateMutex sync.RWMutex
+		state      map[string]State
+
+		chStop chan struct{}
+		chDone chan struct{}
+
+		utils.StartStopOnce
+	}
+
+	State struct {
+		ready   error
+		healthy error
+	}
+
+	Status string
+)
+
+var _ Checker = (*checker)(nil)
+
+const (
+	StatusPassing Status = "passing"
+	StatusFailing Status = "failing"
+
+	interval = 15 * time.Second
+)
+
+var (
+	metrics = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "health",
+			Help: "Health status by service",
+		},
+		[]string{"service_id"},
+	)
+)
+
+func NewChecker() Checker {
+	c := &checker{
+		services: make(map[string]Checkable, 10),
+		state:    make(map[string]State, 10),
+		chStop:   make(chan struct{}),
+		chDone:   make(chan struct{}),
+	}
+
+	return c
+}
+
+func (c *checker) Start() error {
+	return c.StartOnce("HealthCheck", func() error {
+		// update immediately
+		c.update()
+
+		go c.run()
+
+		return nil
+	})
+}
+
+func (c *checker) Close() error {
+	return c.StopOnce("HealthCheck", func() error {
+		close(c.chStop)
+		<-c.chDone
+		return nil
+	})
+}
+
+func (c *checker) run() {
+	defer close(c.chDone)
+
+	ticker := time.NewTicker(interval)
+
+	for {
+		select {
+		case <-ticker.C:
+			c.update()
+		case <-c.chStop:
+			return
+		}
+	}
+
+}
+
+func (c *checker) update() {
+	state := make(map[string]State, len(c.services))
+
+	c.srvMutex.RLock()
+	// copy services into a new map to avoid lock contention while doing checks
+	services := make(map[string]Checkable, len(c.services))
+	for name, s := range c.services {
+		services[name] = s
+	}
+	c.srvMutex.RUnlock()
+
+	// now, do all the checks
+	for name, s := range services {
+		ready := s.Ready()
+		healthy := s.Healthy()
+
+		state[name] = State{ready, healthy}
+	}
+
+	// we use a separate lock to avoid holding the lock over state while talking
+	// to services
```

Comment:
Seems like this can race with a competing thread if it switches context between the previous RLock and this Lock, thereby writing stale data?

Reply:
The health checker runs it's own loop that updates the state, external services only read from it. There's always a current state, when we calculate new state then a short lock is held so we can swap in the new data.